MINI-PROJECTS.md

```diff
@@ -104,6 +104,314 @@ share a bound.
   those bounds is reached. Good, thoughtful test cases are essential
   here!
 
+### Instrumentation to Check Pointer Alignment
+
+Some architectures only allow load instructions / pointer dereferencing when the
+address is a multiple of the word length (i.e. 4 bytes or 8 bytes). This project
+would add a new instrumentation option for `cbmc` and `goto-instrument` that checks
+each pointer before dereference and makes sure it's correctly aligned.
+
+### Command line options to SAT solvers
+
+There are options for which SMT solver to run and there are ways of building with
+multiple SAT solvers but no command line options to pick a SAT solver. It would be
+nice if there were.
+
+César is working on this one. (?)
+
+### Combine Uninitialize Variables Analysis with Alias Analysis
+
+`goto-instrument` features an instrumentation to track cases of use-before-define, i.e.
+reading of uninitialised variables. At present, the analysis does not take into account
+aliasing information. As such, the following code will yield a spurious counterexample:
+
+``` C
+int main()
+{
+  int x;
+  int *ptr;
+  ptr=&x;
+  *ptr=42;
+  if(x < 42)
+    return 1;
+  return 0;
+}
+```
+
+### Constant Folding
+
+With the recently added reaching definitions analysis, it is straightforward to implement
+a program transformation that performs constant folding. ~To ensure soundness in concurrent
+settings, it should be combined with the existing is_threaded analysis.~ (A concurrency-aware
+version of reaching definitions now exists and is awaiting a merge.) Along the way, assertions
+that become trivially true should be eliminated.
```

Comment:
I believe constant folding is actually implemented to the degree described here.

Reply:
`goto-analyzer --simplify out.gb --vsd ...` should do everything except the concurrency aspect.  Good concurrent abstract interpretation is non-trivial.